src/repo/kv.rs

```diff
@@ -0,0 +1,387 @@
+use crate::repo::{PinKind, PinMode, PinStore, References};
+use async_trait::async_trait;
+use cid::{self, Cid};
+use futures::stream::{StreamExt, TryStreamExt};
+use sled::{self, Config as DbConfig, Db, Mode as DbMode};
+use std::convert::Into;
+use std::path::PathBuf;
+use std::str::{self, FromStr};
+use std::sync::Mutex;
+use tracing_futures::Instrument;
+
+#[derive(Debug)]
+pub struct KvDataStore {
+    path: PathBuf,
+    db: Mutex<Option<Db>>,
+}
+
+impl KvDataStore {
+    fn _put(&self, key: &str, value: &str) -> Result<(), Error> {
+        let db = self.get_db();
+
+        let _ = db.insert(key, value)?;
+
+        Ok(())
+    }
+
+    fn _remove(&self, key: &str) -> Result<(), Error> {
+        let db = self.get_db();
+
+        match db.remove(key) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn _apply_batch(&self, batch: sled::Batch) -> Result<(), Error> {
+        let db = self.get_db();
+
+        Ok(db.apply_batch(batch)?)
+    }
+
+    fn get_db(&self) -> Db {
+        self.db.lock().unwrap().as_ref().unwrap().clone()
+    }
+}
+
+#[async_trait]
+impl DataStore for KvDataStore {
+    fn new(root: PathBuf) -> KvDataStore {
+        KvDataStore {
+            path: root,
+            db: Default::default(),
+        }
+    }
+
+    async fn init(&self) -> Result<(), Error> {
+        let config = DbConfig::new();
+
+        let db = config
+            .mode(DbMode::HighThroughput)
+            .path(self.path.as_path())
+            .open()?;
+
+        let mut g = self.db.lock().unwrap();
+        assert!(g.is_none());
+        *g = Some(db);
+
+        Ok(())
+    }
+
+    async fn open(&self) -> Result<(), Error> {
+        Ok(())
+    }
+
+    /// Checks if a key is present in the datastore.
+    async fn contains(&self, _col: Column, _key: &[u8]) -> Result<bool, Error> {
+        Err(anyhow::anyhow!("not implemented"))
+    }
+
+    /// Returns the value associated with a key from the datastore.
+    async fn get(&self, _col: Column, _key: &[u8]) -> Result<Option<Vec<u8>>, Error> {
+        Err(anyhow::anyhow!("not implemented"))
+    }
+
+    /// Puts the value under the key in the datastore.
+    async fn put(&self, _col: Column, _key: &[u8], _value: &[u8]) -> Result<(), Error> {
+        Err(anyhow::anyhow!("not implemented"))
+    }
+
+    /// Removes a key-value pair from the datastore.
+    async fn remove(&self, _col: Column, _key: &[u8]) -> Result<(), Error> {
+        Err(anyhow::anyhow!("not implemented"))
+    }
+
+    /// Wipes the datastore.
+    async fn wipe(&self) {
+        todo!()
+    }
+}
+
+#[async_trait]
+impl PinStore for KvDataStore {
+    async fn is_pinned(&self, block: &Cid) -> Result<bool, Error> {
+        is_pinned(self, block)
+    }
+
+    async fn insert_direct_pin(&self, target: &Cid) -> Result<(), Error> {
+        let already_pinned = get_pinned_mode(self, target)?;
+
+        let mut batch = sled::Batch::default();
```

Comment:
I was thinking if this wasn't potentially costly (in which case conditional creation would be in order), but I've checked and it isn't :+1:.